Fix highlighting lost issues with tab completions and stack pop

pushqrdx opened the pull request:
This should fix #192 #183 #197

For #197 i think it's sufficient to hook `zle-line-init` to `_zsh_highlight`

as for #192 and #183 the issue appears to be gone if i remove `typeset -g _ZSH_HIGHLIGHT_PRIOR_RACTIVE="$REGION_ACTIVE"`
upon further investigation i think that line causes the highlighter to report wrong offsets (only highlights zero chars at the end) hence the color loss, I don't know what is the root cause for that but the edits here are minimal and seem to have fixed all the highlighting issues i faced

FreddieOliveira commented:
I believe this PR also fixes #189. I've pinged its author so he can confirm it.

psprint commented:
The code block should be wrapped with a check for enough Zsh version, as `add-zle-hook-widget` is fairly a recent addition to Zsh.

pushqrdx commented:
@psprint added a check for 5.8

psprint commented:
Could you change the version to 5.4?

pushqrdx commented:
@psprint i have updated the version to `5.4` i also added a fallback for older versions that just uses the old hook method, it could be that we just use the old method and just get rid of `add-zle-hook-widget` but i figured that the `_zsh_highlight_bind_widgets` will eventually migrate to `add-zle-hook-widget` anyways. However one note here, the old binding method is actually needed because i tried to to use `add-zle-hook-widget` for all the bindings and it resulted in a conflict with `zsh-autosuggestions` probably because it didn't properly hook it's functions.

psprint commented:
Thanks.

psprint commented:
Ahso, thanks for the fallback method.
